### PR TITLE
fix: handle unset local value of global-local option

### LINKED ISSUE #15587: vim.bo.undolevels wrong value, most likely underflow?
### Neovim version (nvim -v)

v0.6.0-dev

### Vim (not Nvim) behaves the same?

N.A.

### Operating system/version

ubuntu / mac

### Terminal name/version

iTerm2

### $TERM environment variable

xterm-256color

### Installation

ubuntu: unstable repo, mac: homebrew stable

### How to reproduce the issue

1. open nvim

2. `:edit` a file

3. `lua print(vim.bo.undolevels)`

### Expected behavior

the number printed should be either the global undo level or according to docs:

```txt
	The local value is set to -123456 when the global value is to be used.
```

### Actual behavior

a very large number is printed, different depending on either macos or ubuntu.

I suspect this is a case of underflow happening.

---

I discovered this when trying to set `vim.bo.undolevels = vim.bo.undolevels` to break undo sequences

### PR DESCRIPTION
**Pull request by gpanders**
When `nvim_win_get_option` or `nvim_buf_get_option` are used to get the local value of a global-local number option, and the local value is unset, the return value is an uninitialized number.

For string-valued global-local options, an error is returned. Extend this behavior for number-valued global-local options as well, instead of returning a garbage value. This also makes the Lua `vim.opt` / `vim.o` API work as expected, since those functions depend on `nvim_{but,win}_get_option` returning an error when the local value is unset.

Closes #15587.

### CODE REVIEW

**Comment by dstein64** [comment]
This PR also relates to Issue #13964, in addition to #15587 mentioned in the description.

My expectation, when filing #13964, was that `nvim_win_get_option(0, 'scrolloff')` would match the behavior of `setlocal scrolloff?`. The latter returns `-1` if the local value is unset, whereas the update in this PR would have `nvim_win_get_option` raise an error.

Returning a value instead of raising an error would be consistent with, e.g., what's described in the help for `'undolevels'` ("The local value is set to -123456 when the global value is to be used").

I know there are existing differences between `nvim_win_get_option` and `setl`. For example, on global options the former would error while the latter would return the global value.

@gpanders, did you consider matching the behavior for `setlocal` for the global-local numeric options? If so, what was the motivation for deciding to use the error-based approach?

**Comment by gpanders** [comment]
@dstein64 Sorry for the delayed response, I've been a little side tracked with the diagnostic stuff.

As you know (from other issues you've filed) there is some inconsistency between the Lua options API and the VimL options API (#14670). How this should be resolved is not yet determined, but in the meantime this PR just fixes a more immediate issue of returning an uninitialized value (which is undefined behavior).

File: test/functional/options/num_options_spec.lua [hunk 1]
```diff
@@ -36,6 +37,14 @@ describe(':setlocal', function()
     feed_command('setlocal imsearch=1')
     eq(-1, meths.get_option('imsearch'))
   end)
+
+  it('returns an error if local option does not exist #15587', function()
+    feed_command('setglobal undolevels=1000')
+    feed_command('setlocal undolevels<')
+    eq("Failed to get value for option 'undolevels'",
```
Comment (bfredl) [inline test/functional/options/num_options_spec.lua:44]: IMHO this should also return the sentinel value (not sure why vim choose -123456 for this option and not -2 or something simple but that ship has probably sailed; most global-local options have "sane" sentinels anyway). An unset local value is not an error and thus should not be represented as an error.

Comment (gpanders) [reply test/functional/options/num_options_spec.lua:44]: > An unset local value is not an error and thus should not be represented as an error.

Agreed -- this PR is just a stopgap to address the *immediate* problem of returning an uninitialized value. Fixing the problem more generally still needs to be done. We can just adapt this PR to do that if we don't want to use an intermediate band-aid solution.

Comment (bfredl) [reply test/functional/options/num_options_spec.lua:44]: I just see no reason to implement some code/behaviour we don't want before the code we actually want. Given that we just want the _same_ behaviour as the vimL interfaces I'd assume it is equally small change in `get_option_value_strict` to make that thappen, but, I could of course be wrong.

Comment (bfredl) [reply test/functional/options/num_options_spec.lua:44]: in particular,
```diff
--- a/src/nvim/option.c
+++ b/src/nvim/option.c
@@ -4870,7 +4870,7 @@ int get_option_value_strict(char *name, int64_t *numval, char **stringval, int o
         // only getting a pointer, no need to use aucmd_prepbuf()
         curbuf = (buf_T *)from;
         curwin->w_buffer = curbuf;
-        varp = get_varp(p);
+        varp = get_varp_scope(p, OPT_LOCAL);
         curbuf = save_curbuf;
         curwin->w_buffer = curbuf;
       }
@@ -4878,7 +4878,7 @@ int get_option_value_strict(char *name, int64_t *numval, char **stringval, int o
       win_T *save_curwin = curwin;
       curwin = (win_T *)from;
       curbuf = curwin->w_buffer;
-      varp = get_varp(p);
+      varp = get_varp_scope(p, OPT_LOCAL);
       curwin = save_curwin;
       curbuf = curwin->w_buffer;
     }

```

Comment (gpanders) [reply test/functional/options/num_options_spec.lua:44]: Yep I actually tried that in https://github.com/neovim/neovim/commit/e350feb13bdc8cd9ac941e7284ffc6f2021eeaec but IIRC it broke some other things so I reverted it.

> I just see no reason to implement some code/behaviour we don't want before the code we actually want.

I'll look into this deeper and figure out the right solution.

Comment (bfredl) [reply test/functional/options/num_options_spec.lua:44]: >  it broke some other things so I reverted it.

What things did it break?

Comment (gpanders) [reply test/functional/options/num_options_spec.lua:44]: Nothing insurmountable, it just caused some test failures: https://github.com/neovim/neovim/runs/3556434065

I didn't look into it too deeply at the time, I'll take another look when I get a chance.

Comment (bfredl) [reply test/functional/options/num_options_spec.lua:44]: Well, it _is_ a change of behavior, so test failures are expected to be seen, and not an indication of the change being "wrong" (by necessity)

**Pull request closed without merging**